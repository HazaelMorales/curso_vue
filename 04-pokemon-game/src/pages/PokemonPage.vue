<template>
    <div>
        <h1 v-if ="!pokemon">Cargando...</h1>
        <div v-else>
            <h1>¿Quién es este pokemon?</h1>
            <!--TODO: img-->
            <PokemonPicture 
                :pokemonId="pokemon.id" 
                :showPokemon="showPokemon"/>

            <!--TODO: opciones-->
            <PokemonOptions 
                :pokemons="pokemonArr"
                @selectionPokemon="checkAnswer"/>
            
            <template v-if="showAnswer">
                <h2 class="fade-in">{{ message }}</h2>
                <button @click="newGame">
                    Nuevo Juego
                </button>
            </template>
        </div>
    </div>
</template>

<script>
import PokemonOptions from '@/components/PokemonOptions'
import PokemonPicture from '@/components/PokemonPicture'
import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    components:{PokemonOptions, PokemonPicture},
    data(){
        return{
            pokemonArr:[],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods:{
        async mixPokemonArray(){
            this.pokemonArr = await getPokemonOptions()

            const rndInt = Math.floor(Math.random() * 4)
            this.pokemon = this.pokemonArr[ rndInt ]
        },
        checkAnswer(selectedId){
            this.showPokemon = true
            this.showAnswer = true

            if(selectedId === this.pokemon.id){
                this.message = `Correcto, ${this.pokemon.name}`
            }else{
                this.message = `Oops!, era ${this.pokemon.name}`
            }
        },
        newGame(){
            this.showAnswer = false
            this.showPokemon = false
            this.pokemonArr = []
            this.pokemon = null
            this.mixPokemonArray()
            
        }
    },
    mounted(){
        this.mixPokemonArray()
    }
}
</script>