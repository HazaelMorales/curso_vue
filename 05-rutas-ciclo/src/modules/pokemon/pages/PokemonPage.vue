<template>
    <div>
        <h1>Pokemon Page</h1>
        <h2>Pokemon: <span>#{{ id }}</span></h2>

        <!-- Muestra la imagen del Pokémon si está disponible -->
        <div v-if="pokemon">
            <img :src="pokemon.sprites.front_default" :alt="pokemon.name">
        </div>
        <!-- <img :src="pokemon.sprites.front_default" :alt="pokemon.name"> -->
    </div>
</template>

<script>
export default {
    // Propiedades del componente: 'id' que representa el número de Pokémon
    props: {
        id:{
            type: Number,
            required: true
        },
    },

    // Datos del componente: 'pokemon' que almacena los detalles del Pokémon
    data(){
        return {
            // id : null
            pokemon: null
        }
    },

    // Método llamado cuando se crea el componente para obtener los detalles del Pokémon
    created(){
        // const { id } = this.$route.params
        // console.log(id)
        // this.id = id
        this.getPokemon()
    },
    
    methods:{
        // Método asincrónico para obtener los detalles del Pokémon desde la API
        async getPokemon(){
            try{
                // Realiza una solicitud a la API de Pokémon para obtener los detalles del Pokémon con el id proporcionado
                const pokemon = await fetch(`https://pokeapi.co/api/v2/pokemon/${this.id}`).then(response => response.json())
                console.log(pokemon)
                this.pokemon = pokemon 

            }catch(error){
                // Si ocurre un error al obtener los detalles del Pokémon, redirige al usuario a la página de inicio
                this.$router.push('/')
                console.log('No hay nada que hacer aqui')
            }
        }
    },

    // Vigila cambios en la propiedad 'id' para obtener los detalles del nuevo Pokémon cuando cambia el id
    watch:{
        id(){
            this.getPokemon()
        }
    }
}
</script>
