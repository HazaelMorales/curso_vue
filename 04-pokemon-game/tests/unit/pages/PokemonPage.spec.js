import { shallowMount } from "@vue/test-utils";
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from "../mocks/pokemons.mock";

describe('PokemonPage Component', ()=>{
    let wrapper
    let mixPokemonArraySpy
    let wrapperShallowProps
    
    beforeEach(()=>{
        mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray').mockResolvedValue(pokemons)

        wrapper = shallowMount(PokemonPage)

        wrapperShallowProps = shallowMount(PokemonPage,{
            data(){
                return{
                    pokemonArr:pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })
    })

    test('debe de hacer un match con el snapshot',()=>{
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe de llamar mixPokemonArray al montar', () =>{
        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })

    test('debe de hacer match con el snapshot una vez cargado los pokemons', ()=>{
        expect(wrapperShallowProps.html()).toMatchSnapshot()
    })

    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions',()=>{
        
        //PokemonPicture debe de existir
        //PokemonOptions debe de existir
        const pkpicture = wrapperShallowProps.find('pokemon-picture-stub')
        const pkoption = wrapperShallowProps.find('pokemon-options-stub')

        expect(pkpicture.exists()).toBeTruthy()
        expect(pkoption.exists()).toBeTruthy()

        
        //PokemonPicture attribute pokemonid === 5
        expect(pkpicture.attributes('pokemonid')).toBe('5')


        //PokemonOptions attribute pokemons toBe true
        expect(pkoption.attributes('pokemons')).toBeTruthy()
    })

    test('Prueba con checkAnswer', async()=>{
        await wrapperShallowProps.vm.checkAnswer(5)

        expect(wrapperShallowProps.find('h2').exists()).toBeTruthy()
        expect(wrapperShallowProps.vm.showPokemon).toBeTruthy()
        expect(wrapperShallowProps.find('h2').text()).toBe(`Correcto, ${pokemons[0].name}`)

        await wrapperShallowProps.vm.checkAnswer(10)
        expect(wrapperShallowProps.vm.message).toBe(`Oops!, era ${pokemons[0].name}`)

    })
})