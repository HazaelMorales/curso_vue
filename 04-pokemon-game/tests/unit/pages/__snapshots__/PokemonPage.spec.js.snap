// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot una vez cargado los pokemons 1`] = `
<div>
  <div>
    <h1>¿Quién es este pokemon?</h1>
    <!--TODO: img-->
    <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
    <!--TODO: opciones-->
    <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
    <!--v-if-->
  </div>
</div>
`;

exports[`PokemonPage Component debe de hacer un match con el snapshot 1`] = `
<div>
  <h1>Cargando...</h1>
</div>
`;
